### PR TITLE
[Feature] Implement __eq__ for TensorSpec

### PR DESCRIPTION
**Pull request by omikad**
## Description

This pull request has the following changes:
* Add tests to cover equality check for all the `TensorSpec` subclasses. Equality function returns True if all the fields are the same and class is the same.
* Since majority of subclasses and their dependencies are marked with `@dataclass` - there was no need to implement `__eq__` in many of them. So this PR only adds `__eq__` implementation to a few exceptions.
* Add `__eq__` to the class `ContinuousBox`. Default `__eq__` implementation doesn't work as expected for multi-dimensional tensor comparison. If you have two tensors with more than 1 value then their equality returns not a boolean, but a tensor of the same dimension, which raises "Truth value is ambiguous" error. This PR uses `torch.equal` method instead
* Add `__eq__` to the class `OneHotDiscreteTensorSpec`, since it has an additional attribute `use_register`, and it is not covered by default `__eq__` method provided by `@dataclass` attribute
* Add `__eq__` to the class `CompositeSpec` since it has no attribute `@dataclass`.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.

This change solves #351 issue.

### CODE REVIEW

File: test/test_tensor_spec.py [hunk 1]
```diff
@@ -383,6 +383,302 @@ def test_nested_composite_spec(self, is_complete, device, dtype):
...
+    def _ts_make_all_fields_equal(ts_to, ts_from):
+        ts_to.shape = ts_from.shape
+        ts_to.space = ts_from.space
+        ts_to.device = ts_from.device
+        ts_to.dtype = ts_from.dtype
+        ts_to.domain = ts_from.domain
+        return ts_to
+
+    def test_equality_bounded(self):
+        MIN = 10
```
Comment (vmoens) [inline test/test_tensor_spec.py:397]: can we rename the caps variables in non caps?
we usually use that for global variables

**Review by vmoens — APPROVED** [review]
LGreatTM! 
Thanks for this, it's really useful!
Once the small formatting issue fixed we'll be ready to merge

File: test/test_tensor_spec.py (hunk 1 above)
Comment (omikad) [reply test/test_tensor_spec.py:397]: Sure, one moment

Comment (omikad) [reply test/test_tensor_spec.py]: Done! @vmoens  please check

**Review by vmoens — APPROVED** [review]
Brilliant!

**Pull request merged**